proxy/src/http/project.rs

```diff
@@ -0,0 +1,422 @@
+        state.serialize_field("id", &self.id.to_string())?;
+        state.serialize_field("messages", &self.messages)?;
+        state.serialize_field("state", &self.state)?;
+        state.serialize_field("timestamp", &self.timestamp)?;
+        state.end()
+    }
+}
+
+impl Serialize for registry::Message {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::OrgRegistration(org_id) => serializer.serialize_newtype_variant(
+                "Message",
+                0,
+                "OrgRegistration",
+                &org_id.to_string(),
+            ),
+            Self::OrgUnregistration(org_id) => serializer.serialize_newtype_variant(
+                "Message",
+                1,
+                "OrgUnregistration",
+                &org_id.to_string(),
+            ),
+            Self::ProjectRegistration {
+                org_id,
+                project_name,
+            } => {
+                let mut sv =
+                    serializer.serialize_struct_variant("Message", 2, "ProjectRegistration", 2)?;
+                sv.serialize_field("org_id", &org_id.to_string())?;
+                sv.serialize_field("project_name", &project_name.to_string())?;
+                sv.end()
+            },
+            Self::UserRegistration { handle, id } => {
+                let mut sv =
+                    serializer.serialize_struct_variant("Message", 3, "UserRegistration", 2)?;
+                sv.serialize_field("handle", &handle.to_string())?;
+                sv.serialize_field("id", &id.to_string())?;
+                sv.end()
+            },
+        }
+    }
+}
+
+impl Serialize for registry::TransactionState {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::Applied(block_hash) => {
+                let mut state = serializer.serialize_struct("TransactionApplied", 2)?;
+                state.serialize_field("type", "TransactionApplied")?;
+                state.serialize_field("block_hash", &block_hash.to_string())?;
+
+                state.end()
+            },
+        }
+    }
+}
+
+/// Bundled input data for project creation.
+#[derive(Deserialize, Serialize)]
+pub struct CreateInput {
+    /// Location on the filesystem of the project, an empty directory means we set up a fresh git
+    /// repo at the path before initialising the project.
+    path: String,
+    /// User provided metadata for the project.
+    metadata: MetadataInput,
+}
+
+/// User provided metadata for project manipulation.
+#[derive(Deserialize, Serialize)]
+pub struct MetadataInput {
+    /// Name of the proejct.
+    name: String,
+    /// Long form outline.
+    description: String,
+    /// Configured default branch.
+    default_branch: String,
+    /// Display image of the project.
+    img_url: String,
```

Comment:
don't know the context so forgive-me if it doesn't make sense, but hyper has an [Uri field that impl's From<String> ](https://docs.rs/hyper/0.13.1/hyper/struct.Uri.html#impl-TryFrom%3CString%3E) and has many methods that may be useful while dealing with url's

Reply:
Didn't know about the `Uri` type in hyper, might be useful down the road. But your comment triggered me to enforce our already in use `Url` type in the core methods as it is parse to one anyway. Good pointer!

Reply:
See 76b1737.

Reply:
ah nice, btw I found https://github.com/radicle-dev/radicle-link/blob/master/librad/src/meta/common.rs#L164
nitpick: rather than having a parse method, you can impl either the `FromStr` or `From<String>` traits for Url, std does it [here for example](https://doc.rust-lang.org/src/core/str/mod.rs.html#4119-4121)